tests/EdgeToEdge/Routes/SofortPaymentNotificationRouteTest.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Tests\EdgeToEdge\Routes;
+
+use DateTime;
+use Psr\Log\LogLevel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Client;
+use WMDE\Fundraising\Frontend\DonationContext\Tests\Data\ValidDonation;
+use WMDE\Fundraising\Frontend\Factories\FunFunFactory;
+use WMDE\Fundraising\Frontend\Tests\EdgeToEdge\WebRouteTestCase;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\FixedTokenGenerator;
+use WMDE\PsrLogTestDoubles\LoggerSpy;
+
+class SofortPaymentNotificationRouteTest extends WebRouteTestCase {
+
+	private const TOKEN_VALID = 'my-secret_token';
+	private const TOKEN_INVALID = 'fffffggggg';
+
+	public function testGivenWrongPaymentType_applicationRefuses(): void {
+		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ) {
+			$factory->setTokenGenerator( new FixedTokenGenerator(
+				self::TOKEN_VALID,
+				new DateTime( '2039-12-31 23:59:59Z' )
+			) );
+
+			$repo = $factory->getDonationRepository();
+			$repo->storeDonation( ValidDonation::newIncompletePayPalDonation() );	// creates donation w/ id=1
```

Comment:
Like this: https://github.com/wmde/FundraisingFrontend/pull/923/commits/060f8abcbaab0b610f8dcac8d3af127a0d53c12b